File: python/organisation.py
from typing import Iterable, List

from award import Award

class Organisation:
    # id is renamed to id_ to avoid conflict with the built-in id() function
    def __init__(self, id_: int, issued_awards: Iterable[Award]):
        self._id: int = id_
        self._issued_awards: List[Award] = list(issued_awards)

    def get_id(self) -> int:
        return self._id

    def get_issued_awards(self) -> Iterable[Award]:
        return self._issued_awards

    def check_one_award_per_game_per_year(self) -> bool:
        # issuedAwards->forAll(a1, a2 | ...)
        for a1 in self._issued_awards:
            for a2 in self._issued_awards:
                # (a1 <> a2 and a1.forGame = a2.forGame)
                premise = (a1 is not a2) and (a1.get_for_game() is a2.get_for_game())

                # implies a1.at.year() <> a2.at.year()
                if premise:
                    conclusion = a1.get_date().year != a2.get_date().year

                    # implies: if premise is true, conclusion must be true
                    if not conclusion:
                        return False
        return True

File: java/Organisation.java
import java.util.LinkedList;
import java.util.List;

public class Organisation {
    private int id;
    private List<Award> issuedAwards = new LinkedList<>();

    public Organisation(int id, Iterable<Award> issuedAwards) {
        this.id = id;
        for (Award award : issuedAwards) {
            this.issuedAwards.add(award);
        }
    }

    public int getId() { return id; }

    public Iterable<Award> getIssuedAwards() { return issuedAwards; }

    public boolean checkOneAwardPerGamePerYear() {
        // issuedAwards->forAll(a1, a2 | ...)
        for (Award a1 : issuedAwards) {
            for (Award a2 : issuedAwards) {
                // (a1 <> a2 and a1.forGame = a2.forGame)
                boolean premise = (a1 != a2) && 
                    (a1.getForGame() == a2.getForGame());
                
                // implies a1.at.year() <> a2.at.year()
                if (premise) {
                    boolean conclusion = 
                        a1.getDate().getYear() != a2.getDate().getYear();
                    
                    // implies: if premise is true, conclusion must be true
                    if (!conclusion) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
}
